### ISSUE TITLE
Google Earth Engine mini Summit 2019-05-10 TensorFlow with EE

### ISSUE THREAD

**Issue by mapconcierge** [furuhashilab/GEE4GSC#1]
# TensorFlow with EE
http://bit.ly/TF_TOK_mini_2019

## 準備
* Google Colaboratory を使います。
https://colab.research.google.com/notebooks/welcome.ipynb

## What this course IS about
* An end-to-end workflow for using TensorFlow to perform classification on data from Earth Engine.
* A crash course in the TensorFlow Keras API
* An introduction to Colab

<img width="1041" alt="スクリーンショット 2019-05-10 11 03 38" src="https://user-images.githubusercontent.com/416977/57497605-51797d00-7313-11e9-87f6-77817289fbf7.png">

## 概念
* 
<img width="1019" alt="スクリーンショット 2019-05-10 14 23 29" src="https://user-images.githubusercontent.com/416977/57504396-3a488880-732f-11e9-9a30-a3a11803d055.png">

* 
<img width="1040" alt="スクリーンショット 2019-05-10 14 23 40" src="https://user-images.githubusercontent.com/416977/57504398-3ddc0f80-732f-11e9-8a5d-2f4a85d848b8.png">

* 出力されるのは Probability とタイプ
<img width="1035" alt="スクリーンショット 2019-05-10 11 11 45" src="https://user-images.githubusercontent.com/416977/57497863-6dc9e980-7314-11e9-8678-6670ce771cdb.png">

* TensorFlow ガイド
https://www.tensorflow.org/guide/premade_estimators#create_input_functions

## 実行環境
* 以下のNoteを開き自分の Google Drive にコピーする。
https://colab.research.google.com/github/google/earthengine-api/blob/master/python/examples/ipynb/TF_demo1_keras.ipynb

* コピーしたnoteで Install the Earth Engine client library を実行する。

* Test the Folium installation の Folium は Mapping library . Leaflet? と言ってた。→ Leaflet でした :-)
<img width="1063" alt="スクリーンショット 2019-05-10 11 25 20" src="https://user-images.githubusercontent.com/416977/57498335-6dcae900-7316-11e9-85a4-df232be9ff52.png">

<img width="1440" alt="スクリーンショット 2019-05-10 11 51 12" src="https://user-images.githubusercontent.com/416977/57499256-057e0680-731a-11e9-8329-ad7a6aae7bd8.png">

`
import folium
print(folium.__version__)

# Define the URL format used for Earth Engine generated map tiles.
EE_TILES = 'https://earthengine.googleapis.com/map/{mapid}/{{z}}/{{x}}/{{y}}?token={token}'
`

# outputBucket のグループを自分用に変更する。
outputBucket = 'ee-docs-demos'
outputBucket = 'nclinton-training-temp'

個人用に　mapconcierge というBucketを Google Cloud Platform 上に作ってみた。
<img width="809" alt="スクリーンショット 2019-05-10 11 44 53" src="https://user-images.githubusercontent.com/416977/57499069-3c075180-7319-11e9-843d-b47a2d7b5329.png">

<img width="1051" alt="スクリーンショット 2019-05-10 12 08 15" src="https://user-images.githubusercontent.com/416977/57499914-55f66380-731c-11e9-96a0-44bb1c392397.png">

<img width="1050" alt="スクリーンショット 2019-05-10 12 09 07" src="https://user-images.githubusercontent.com/416977/57499952-74f4f580-731c-11e9-95a6-d57201ba93ef.png">

* Parse the dataset
<img width="1050" alt="スクリーンショット 2019-05-10 12 10 12" src="https://user-images.githubusercontent.com/416977/57500002-a2da3a00-731c-11e9-8a22-41d7676aa6d2.png">

<img width="1049" alt="スクリーンショット 2019-05-10 12 10 30" src="https://user-images.githubusercontent.com/416977/57500006-a8d01b00-731c-11e9-89a1-0073ab6846b8.png">

<img width="1090" alt="スクリーンショット 2019-05-10 12 15 43" src="https://user-images.githubusercontent.com/416977/57500308-eda88180-731d-11e9-95b4-7e76afbf9c8f.png">

* Check model accuracy on the test set. 最後の "1.0" が 精度100%を意味する。

19/100 [====>.........................] - ETA: 1s - loss: 0.2697 - acc: 1.0000 WARNING:tensorflow:Your dataset iterator ran out of data.
[0.0773512352257967, **1.0**]

<img width="1089" alt="スクリーンショット 2019-05-10 12 20 21" src="https://user-images.githubusercontent.com/416977/57500466-7de6c680-731e-11e9-9217-8cd2b0026015.png">

<img width="1090" alt="スクリーンショット 2019-05-10 12 21 01" src="https://user-images.githubusercontent.com/416977/57500476-850dd480-731e-11e9-82cc-3cce3294dcc1.png">

<img width="1091" alt="スクリーンショット 2019-05-10 12 20 43" src="https://user-images.githubusercontent.com/416977/57500473-80e1b700-731e-11e9-843f-5e7a601340b3.png">

* Generate predictions for the image pixels はかなり時間かかる。（サンプルコードで約10分）
<img width="1092" alt="スクリーンショット 2019-05-10 12 27 52" src="https://user-images.githubusercontent.com/416977/57500644-19783700-731f-11e9-8c18-3821909b2219.png">

<img width="1090" alt="スクリーンショット 2019-05-10 12 31 05" src="https://user-images.githubusercontent.com/416977/57500765-9b686000-731f-11e9-9bf4-bf30a14e8fa2.png">

`130/130` [==============================] - 524s 4s/step
[[0.09496748 0.8952058  `0.00982673]]`

### ここまでの Bucket 消費状況
<img width="1165" alt="スクリーンショット 2019-05-10 12 23 19" src="https://user-images.githubusercontent.com/416977/57500464-79baa900-731e-11e9-9e0b-2f30c31a7d16.png">

https://console.cloud.google.com/billing

<img width="1090" alt="スクリーンショット 2019-05-10 12 35 34" src="https://user-images.githubusercontent.com/416977/57500881-32351c80-7320-11e9-90a2-fa759c6d41c6.png">

<img width="1094" alt="スクリーンショット 2019-05-10 12 50 35" src="https://user-images.githubusercontent.com/416977/57501374-4548ec00-7322-11e9-8036-cd6cebac2e33.png">

<img width="1095" alt="スクリーンショット 2019-05-10 12 51 50" src="https://user-images.githubusercontent.com/416977/57501428-66a9d800-7322-11e9-90f8-3b858014b607.png">

ここで、USER_NAME を nclinton のままにしていたことに気づき、修正。
`USER_NAME = 'mapconcierge'
outputAssetID = 'users/' + USER_NAME + '/Classified_pixel_demo'
print('Writing to ' + outputAssetID)`

<img width="1094" alt="スクリーンショット 2019-05-10 12 54 42" src="https://user-images.githubusercontent.com/416977/57501708-7f66bd80-7323-11e9-94ec-7f9c1bc43304.png">

<img width="1094" alt="スクリーンショット 2019-05-10 12 54 54" src="https://user-images.githubusercontent.com/416977/57501714-8392db00-7323-11e9-8d5c-8b4ee6fb1ae5.png">

画像の出力にかなり時間がかかる。
Google Earth Engine の Code Editor タスクで進捗がわかる。

<img width="328" alt="スクリーンショット 2019-05-10 13 02 53" src="https://user-images.githubusercontent.com/416977/57504119-26e8ed80-732e-11e9-99a0-3b02c89c7bdc.png">

<img width="1127" alt="スクリーンショット 2019-05-10 14 14 04" src="https://user-images.githubusercontent.com/416977/57504136-3f590800-732e-11e9-83bc-bf1f0c711b6c.png">

最終的な Google Cloud Bucket 消費ストレージ量。このデモでは 約200MB 使用。
<img width="1433" alt="スクリーンショット 2019-05-10 14 32 18" src="https://user-images.githubusercontent.com/416977/57504743-9829a000-7330-11e9-9146-caa5d29dd152.png">

## 解析結果を Google Earth Engine に読み込んで処理は終了。
* Tasks の ? から View Assets 経由でデータ読み込み。
<img width="1439" alt="スクリーンショット 2019-05-10 16 06 16" src="https://user-images.githubusercontent.com/416977/57508928-d5485f00-733d-11e9-90e6-7a9aa3a75141.png">

<img width="701" alt="スクリーンショット 2019-05-10 16 07 20" src="https://user-images.githubusercontent.com/416977/57508934-da0d1300-733d-11e9-906a-f018c84a963d.png">

<img width="1440" alt="スクリーンショット 2019-05-10 16 11 20" src="https://user-images.githubusercontent.com/416977/57509084-4ab42f80-733e-11e9-8a60-830a2537844b.png">

**Comment by mapconcierge** [comment]
# TIPs

## TensorFlow の Keras API について知っておく必要がある。
https://qiita.com/cometscome_phys/items/d9553fe7c92e09fc14a9

## ipynb(Jupyter Notebook) GitHub上で公開することができる。今回のサンプルコードも同様。
http://pppurple.hatenablog.com/entry/2016/06/11/194713

## エクスクラメーションマーク ! は bash コマンドであることを意識する。
ex. !earthengine authenticate 

## Python の Folium ライブラリは重要！
https://python-visualization.github.io/folium/

## XYZタイルへのアクセス方法 Define the URL format used for Earth Engine generated map tiles.
EE_TILES = 'https://earthengine.googleapis.com/map/{mapid}/{{z}}/{{x}}/{{y}}?token={token}'

## epoch の概念
https://www.st-hakky-blog.com/entry/2017/01/17/165137

## TensorFlow で DatasetをZIP圧縮で扱う方法
https://www.tensorflow.org/api_docs/python/tf/data/Dataset#zip

## gsutil コマンドは知っておくべき
https://www.apps-gcp.com/gsutil-command-explanation-part1/

## argmax 配列の最大要素のインデックスを返す関数
https://deepage.net/features/numpy-argmax.html

## Google Cloud Storage のBucket詳細でRollも変更可能
<img width="713" alt="スクリーンショット 2019-05-10 15 02 41" src="https://user-images.githubusercontent.com/416977/57505857-fc4e6300-7334-11e9-8e04-5301db6800dc.png">

## softmax関数についても知っておく
http://hiro2o2.hatenablog.jp/entry/2016/07/21/013805
